### ISSUE TITLE
Libtorch torch::cat fails with C++ memory exception after upgrading to 1.2 stable

### ISSUE THREAD

**Issue by Alex18947** [pytorch/pytorch#25457]
## 🐛 Bug

Using Pytorch 1.2 stable on Windows, a simple torch::cat of two GPU tensors generates a C++ memory exception. With CPU tensors, no problem happens.

## To Reproduce

Run this code to reproduce in a C++ console:

	auto t1 = torch::rand({ 2, 2 });
	auto t2 = torch::rand({ 2, 2 });
	t1 = t1.to(torch::kCUDA);
	t2 = t2.to(torch::kCUDA);
	auto t3 = torch::cat({ t1, t2 });
	cout << t3;

You will observe a C++ memory exception in 

`static inline Tensor cat(TensorList tensors, int64_t dim)`

around globalATenDispatch.

## Expected behavior

The concatenated tensor should be output to the console.

## Environment

 - PyTorch 1.2
 - Win 10
 - Installed from source
 - Built with Ninja
 - CUDA 10.1
 - GTX 1060

cc @peterjc123

**Comment by peterjc123** [comment]
This seems to be another initialization problem just like #22681.

**Comment by Alex18947** [comment]
In contrast to [22681](https://github.com/pytorch/pytorch/issues/22681) this happens in release mode. It sounds similar to [25298](https://github.com/pytorch/pytorch/issues/25298).

I honestly don't understand how can I be the first one to bring this out - seems like nobody is using Windows anymore ;). With these basis tensor ops failing, the entire 1.2 framework is just unusable. Maybe it's time to stop Windows support at all, or somehow make sure that Windows users can use the framework as well.

**Comment by xsacha** [comment]
I am using 1.2 on Windows :) In production. Using torch::cat for CUDA tensors hundreds of times a second as well.
There must be some other condition triggering this one.

**Comment by Alex18947** [comment]
For me, torch::NoGradGuard makes no difference. And yes, some other condition must be triggering this, in #25298 the same problem with globalATenDispatch happens when copying to GPU, so in a different place.

**Comment by soumith** [comment]
@yf225 @peterjc123 we run a full suite of Python+Windows tests on every CI commit, but have we enabled them for libtorch / C++ frontend? If not, can we please enable them asap, so that something like this and #25298 cannot repeat. Please let me know if I misunderstood the context.

**Comment by soumith** [comment]
okay, I spoke to @peterjc123 and we actually have libtorch tests enabled in the CI, and everything described in this post works on Windows for our CI version of VS / WinSDK.

> I honestly don't understand how can I be the first one to bring this out - seems like nobody is using Windows anymore ;). With these basis tensor ops failing, the entire 1.2 framework is just unusable. Maybe it's time to stop Windows support at all, or somehow make sure that Windows users can use the framework as well.

I wish you maybe asked a bit more nicely, instead of coming to some kind of drastic conclusion. Developers like @peterjc123 maintain PyTorch for Windows in their free time and put a lot of time and energy to help make the project better, and ignorant comments such as yours hurts them deeply. The tests pass on the CI but fail at your end (as you see Sacha had also said they work on his end). It's possible that they are failing on a particular version of VS or Windows SDK, and giving more details will help.

**Comment by Alex18947** [comment]
Well, after reading #25298 which has the exact same symptoms I was under the impression that this is a general problem. So now it's clear that it is not that easy, so I will try to be less ignorant and investigate further on my side. I am sure that #25298 and my problems are actually of the same nature.

**Comment by xsacha** [comment]
Did you try avoiding the GPU copy?
As:
```C++
auto t1 = torch::rand({ 2, 2 }, torch::kCUDA);
auto t2 = torch::rand({ 2, 2 }, torch::kCUDA);
auto t3 = torch::cat({ t1, t2 })
```

I never copy in my code. I leave everything on the GPU it came on. That could be the cause of all those issues.

**Comment by Alex18947** [comment]
Ok, so here is the story. Maybe it helps others experiencing these C++ memory exceptions (e.g. #25298) Looking closer at libtorch 1.1 vs. 1.2 prebuilt binaries, I now see that there are 3 new nvrtc libraries added (runtime compilation). Since I was always building from source, I failed to add these libraries to the final application folder. I checked, of course, the torch.dll 1.2 dependencies (whether there are some new dlls I should add), but torch.dll does not directly depend on caffe2_nvrtc.dll. That's why Windows didn't report the usual missing dll problem (which would be easy). Instead, it fails with some weird memory C++ exceptions along the new dispatcher code.

Thanks for pointing me into the right direction (i.e. my ignorance) and keep up the good work ;).

PS. Why is nvrtc needed now?
